@MainActor on protocols with associated types causes compiler error: Circular reference

**Describe the bug**
This code produces the compiler error: Circular reference, 
```swift
@MainActor protocol ProtocolWithAssociatedTypes {
    associatedtype ID: Hashable
    associatedtype Value

    subscript(_ id: ID) -> Value { get set }
}

final class ClassConforming<ID: Hashable,Value>: ProtocolWithAssociatedTypes {
    subscript(id: ID) -> Value {
        get { fatalError() }
        set { fatalError() }
    }
}
```
removing `@MainActor` makes it compile 
```swift
protocol ProtocolWithAssociatedTypes {
```

**Steps To Reproduce**
Steps to reproduce the behavior:
1. type the code from above
2. swift build


**Expected behavior**
Should not be an error

**Screenshots**
```
Building for debugging...
/Generic_MainActor_Bug.swift:9:5: error: circular reference
    subscript(id: ID) -> Value {
    ^
/Generic_MainActor_Bug.swift:8:13: note: through reference here
final class ClassConforming<ID: Hashable,Value>: ProtocolWithAssociatedTypes {
            ^
/Generic_MainActor_Bug.swift:9:5: error: circular reference
    subscript(id: ID) -> Value {
    ^
/Generic_MainActor_Bug.swift:8:13: note: through reference here
final class ClassConforming<ID: Hashable,Value>: ProtocolWithAssociatedTypes {
```

**Environment (please fill out the following information)**
 - OS: macOS 13.0
 - ```swift-driver version: 1.62.15 Apple Swift version 5.7.1 (swiftlang-5.7.1.134.3 clang-1400.0.29.51)```
 - ```swift-driver version: 1.62.8 Apple Swift version 5.7 (swiftlang-5.7.0.127.4 clang-1400.0.29.50)```

**Additional Info**

removing `subscript` also compiles
```swift
@MainActor protocol ProtocolWithAssociatedTypes {
    associatedtype ID: Hashable
    associatedtype Value

    func getValue(id: ID) -> Value
    func setValue(_ value: Value, id: ID)
}

@MainActor final class ClassConforming<ID: Hashable,Value>: ProtocolWithAssociatedTypes {
    func getValue(id: ID) -> Value {
        fatalError()
    }

    func setValue(_ value: Value, id: ID) {
        fatalError()
    }
}
```

--- a/lib/Sema/ConstraintSystem.cpp
+++ b/lib/Sema/ConstraintSystem.cpp
@@ -2427,8 +2427,9 @@ ConstraintSystem::getTypeOfMemberReference(
 
   // Adjust the opened type for concurrency.
   Type origOpenedType = openedType;
-  if ((isa<AbstractFunctionDecl>(value) || isa<EnumElementDecl>(value)) &&
-      !isRequirementOrWitness(locator)) {
+  if (isRequirementOrWitness(locator)) {
+    // Don't adjust when doing witness matching, because that can cause cycles.
+  } else if (isa<AbstractFunctionDecl>(value) || isa<EnumElementDecl>(value)) {
     unsigned numApplies = getNumApplications(
         value, hasAppliedSelf, functionRefKind);
     openedType = adjustFunctionTypeForConcurrency(

--- a/test/Concurrency/actor_isolation_cycle.swift
+++ b/test/Concurrency/actor_isolation_cycle.swift
@@ -10,3 +10,20 @@ public struct S : P {
   public func g(_: Int) {}
   public func f(_: T) {}
 }
+
+// https://github.com/apple/swift/issues/61602
+@available(SwiftStdlib 5.1, *)
+@MainActor protocol ProtocolWithAssociatedTypes {
+    associatedtype ID: Hashable
+    associatedtype Value
+
+    subscript(_ id: ID) -> Value { get set }
+}
+
+@available(SwiftStdlib 5.1, *)
+final class ClassConforming<ID: Hashable,Value>: ProtocolWithAssociatedTypes {
+    subscript(id: ID) -> Value {
+        get { fatalError() }
+        set { fatalError() }
+    }
+}

